[yeon] 미션5: 보드에 위치 부여 및 점수계산

kimnayeon0108 opened the pull request:
안녕하세요 백엔드반 yeon입니다! 
이번 미션은 이전 것들보다 시간도 오래걸리고 고민도 훨씬 많이 한 시간이었습니다.
구현하면서 고민했던 부분에 대해서 설명드리겠습니다.


**Board 클래스**
1. 요구사항에 나와있는대로 createWhite(type)과 createBlack(type)메소드를 생성하고, 각각의 기물들의 팩토리메소드에서 이를 호출하도록 구현했습니다.
그런데  각각 기물을 생성하는 메소드가 굳이 필요한가? 기물생성시 createWhite()나 createBlack()을 써서 생성하면 안되나 라는 생각이 들었지만, 팩토리메소드의 특징상 메소드명에 기물도 포함되어야할거 같다는 생각도 들었습니다.

2. 남아있는 기물의 점수 합산하기
같은 세로줄에 있는 폰은 0.5점으로 계산해야하는것을 구현하기 위해 getSameFileNum()으로 같은 세로줄에 있는 폰의 개수를 구하는 메소드를 생성했습니다. 
이 메소드를 구현하는데에 제일 많이 고민했습니다. 여러가지를 고민하다가 HashSet은 요소의 중복이 없다는 것을 이용하여 폰의 위치를 담은 배열(positions)을 HashSet으로 바꾸고, 크기의 차이를 이용해서 값을 구하는 것이 코드가 제일 깔끔한것 같아서 이렇게 구현하기로 결정했습니다.
근데 보는 사람 입장에서는 갑자기 HashSet이 왜나왔지? 라는 생각이 들수도 있을거같습니다. 

사실 더 많은 고민사항들이 있었지만 가장 많이 고민한 부분만 정리해보았습니다.
바쁘신데 리뷰해주셔서 감사합니다!!!🙇‍♀️🙇‍♀️

kimnayeon0108 committed:
Move enum Color, PieceName into Piece class
rename enum PieceName to Type

kimnayeon0108 committed:
Add NOCOLOR in enum Color
add NO_PIECE in enum Type

kimnayeon0108 committed:
Add getRepresentation() method
기물의 종류에 따른 representation이 잘 지정되는지 확인한다.

kimnayeon0108 committed:
Add getBlackRepresentation(), getWhiteRepresentation() method in enum Type
기존의 tempRepresentation을 representation으로 변수명을 바꾸고 각각 색깔에 맞는 문자를 반환하도록 메소드를 추가했다.
Piece클래스의 createRepresentation의 메소드도 추가한 메소드를 호출하도록 코드를 수정했다.

kimnayeon0108 committed:
Fix createPiece(), verifyPiece() method
기존에 create_piece() 메소드와 verify_piece() 메소드를 수정함
Piece 클래스의 팩토리메소드로 기물을 생성하고 기물의 타입과 색깔이 올바르게 지정되었는지 확인하는 역할을 한다.

kimnayeon0108 committed:
Add createBlank() method
빈 기물을 생성하였을때 color와 타입이 올바르게 지정되었는지 확인하는 기능 추가

kimnayeon0108 committed:
Add createWhite(), createBlack() method
기물들을 생성하는 팩토리메소드들의 중복을 제거하기 위한 메소드 추가
팩토리메소드 안에서 호출하도록 코드 수정

kimnayeon0108 committed:
Add Rank class and fix Board class
체스판의 Rank를 담당하는 Rank 클래스 생성
Board 가 8개의 Rank 객체를 List에 담는 기능을 위해 Rank 클래스 생성함

kimnayeon0108 committed:
Fix Board class
기존의 기물을 담는 리스트를 삭제하고 Rank 객체의 pieceList에 기물을 담도록 구현함

kimnayeon0108 committed:
Add count piece function
Rank 클래스의 pieceList에서 해당하는 기물의 개수를 반환하는 count()메소드를 생성하고 Board 클래스의 countPiece() 메소드에서 이를 호출하여 보드 전체에서의  기물의 개수를 반환하도록 구현

kimnayeon0108 committed:
Add count() test method
기물의 개수를 올바르게 반환하는지 확인하는 테스트 코드 추가

kimnayeon0108 committed:
Add findPiece() method
해당 위치의 기물이 올바르게 반환되는지 확인하는 테스트 코드 작성

kimnayeon0108 committed:
Add findPiece() method
Board클래스에 인자로 받은 위치의 기물을 반환하는 기능 구현

kimnayeon0108 committed:
Overide eauals() and hashCode() in Piece class
기물들을 value object로 만들기 위해 equals()와 hashCode() 추가

kimnayeon0108 committed:
Add setPiece() method in Board class
기물을 해당하는 위치에 배치해주는 메소드 구현
위치의 x,y축을 얻기위한 getX() getY() 메소드 추가

kimnayeon0108 committed:
Add function calculate total point of each color pieces
폰이 같은 세로줄에 있는지 검사하기 위해 isSameFileWhite(), isSameFileBlack() 메소드 추가
보드의 전체 칸을 돌면서 해당 기물의 위치를 배열로 반환하는 findPosition() 메소드 추가, 같은 세로줄에 폰이 있는지를 확인하기 위해서는 위치의 값들이 필요하기때문에 추가한 기능

kimnayeon0108 committed:
Refactor duplicated code
점수의 총합을 구하는 메소드를 색깔별로 구현한 것을 하나의 메소드로 통일해서 인자로 색을 받도록 하기위해 코드 수정함

kimnayeon0108 committed:
Fix calculate point function
폰의 점수를 계산하는 부분을 잘못 구현해서 수정
같은 세로줄에 있는 폰의 개수를 구하는 deletedPositionNum() 메소드로 변경

kimnayeon0108 committed:
Fix getSameFileNum() method
pawn의 위치를 담는 배열을 hashSet으로 변경하면 자동으로 중복이 사라지는걸 이용해서 같은 세로줄에 있는 폰의 개수를 구하도록 코드 수정

kimnayeon0108 committed:
Add sortPiece() method in board class
남아있는 기물을 색깔별로 구분하고 점수의 내림차순으로 정렬한 리스트를 반환하는  메소드 추가

kimnayeon0108 committed:
Refactor code
리팩토링

kimnayeon0108 committed:
코드 리팩토링

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,127 +1,138 @@
 
-    private String getBlackPawnResult() {
-        return getPieceResult(blackPawns);
+        return getSum(pawn, sameFilePawn, knight, rook, bishop, queen);
```
설계의 변경이 좀 필요할 것으로 보이는데요, `Piece` 가 자신의 점수를 갖고 있으면서,
`Type` 에 맞는 `Piece` 를 골라낸 후 `Piece` 를 점수의 형태로 변환해서 그 합계를 구하면 훨씬 깔끔한 구현이 가능할 것이라고 생각합니다. 

_키워드_

- `.stream()`
- `.map()`
- `.sum()`

kimnayeon0108 replied:
@wheejuni 아하 감사합니다! getSum()부분이 매개변수도 너무 많은거 같고 찜찜했는데 말씀해주신대로 고민하고 변경하도록 하겠습니다!

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,127 +1,138 @@
+        return positionArr;
+    }
+
+    private double getSum(double pawn, double sameFilePawn, double knight, double rook, double bishop, double queen) {
```
설계가 개선되면 이 메소드는 삭제될 거라고 생각해요. 
이 메소드 없이 점수의 합계를 구할 수 있는 방법을 고민해주세요.

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -1,127 +1,138 @@
-    private String getBlackPieceResult() {
-        return getPieceResult(blackPieces);
+    private double getSameFileNum(double pawnNum, Color color) {
+        Piece pawn =
+                color == Color.WHITE ? Piece.createWhitePawn() : Piece.createBlackPawn();
```
삼항 연산자를 사용한 건 좋은데요, 문제는 할당 연산자(`=`) 후에 바로 등호(`==`) 가 와서 가독성이 떨어지는 것 같아요.
이런 경우에는 그냥 `if` 문을 통해서 선언해주는 게 나을 수도 있겠습니다.

kimnayeon0108 replied:
@wheejuni 아하 이부분이 코드읽을때 뭔가 이상하다고 생각했는데 if문을 사용하는게 더 낫군요 감사합니다!!!🙇‍♀️👍

wheejuni reviewed (changes requested):
5단계 진행하시느라 고생 많으셨습니다. 💯 
전반적으로 점수 구하는 로직 이외에는 큰 무리 없이 깔끔하게 구현이 잘 되었습니다.
다만 점수 구하기 로직에는 큰 설계상의 변경이 필요할 걸로 생각됩니다. 
이 부분을 먼저 같이 고민해본 후에 추가 리뷰 진행하면 좋겠는데요. 
점수 구하는 로직에 대해서 리팩토링 의견을 조금 드렸는데 참고해주시고 더 나온 로직으로 작성해 주시면 좋겠네요.

kimnayeon0108 committed:
Add Position class
기물의 위치에 대한 정보를 관리하기 위한 Position 클래스 추가

kimnayeon0108 committed:
Add position variable in Piece class
기물이 position 속성을 갖게 하고 기물 생성시 position을 지정하도록 생성자 변경
PieceTest 클래스에 position이 올바르게 지정되는지 확인하는 테스트 코드 추가

kimnayeon0108 committed:
코드 리팩토링

wheejuni commented:
@kimnayeon0108 아직 수정중이신가요?

kimnayeon0108 commented:
@wheejuni 같은 세로줄에 있는 폰 구하는거에서 시간이 오래걸리고 있습니다ㅜㅜ 원래 구현방식에서 다르게 해보는중입니다...!

wheejuni commented:
@kimnayeon0108 네 천천히 하세요.

kimnayeon0108 committed:
Add getting point of pawns on the same file
같은 파일에 있는 폰의 점수를 구하는 getSameFilePawnPoint() 메소드 추가
폰의 개수를 파일별로 담은 map을 반환하는 getPawnFileMap() 메소드 추가
map의 value를 이용하여 중첩된 폰의 개수를 구함

kimnayeon0108 committed:
Remove useless method

kimnayeon0108 committed:
코드 줄맞춤

kimnayeon0108 commented:
@wheejuni 코드 수정 완료했습니다!

변경사항
위치의 정보를 갖는 Position 클래스를 추가하고, Piece의 객체가 point 정보를 갖도록 추가했습니다. 
점수 구현하는 부분은 말씀해주신대로 stream()과 sum()을 이용해서 변경했습니다.

세로줄에 중첩된 폰을 구하는데 groupingBy()를 이용해보았습니다. 여기서 얻은 map의 value를 이용해서 중첩된 폰의 개수를 얻도록 했습니다.

 groupingBy()와 map 자료구조를 사용하는 부분에서 시간이 많이 걸렸습니다. 오래 기다리게해서 죄송합니다ㅜㅜ 처음 써보는 함수와 자료구조라서 어려웠던 탓입니다. 제 머리속에 떠오르는것 중에서 이방법이 제일 적합하다고 생각해서 시도해보았습니다.
너무 복잡하게 구현된거 같고 올바른 방법인지는 모르겠어서 이부분에 대한 피드백 주시면 감사하겠습니다...!!!

wheejuni commented on src/main/java/chess/Board.java:
```diff
@@ -7,26 +7,27 @@
     public void initializeEmpty() {
-        for (int i = 0; i < BOARD_RANK; i++) {
-            board.add(Rank.initBlankRank());
+        for (int rankNum = 1; rankNum <= BOARD_RANK; rankNum++) {
```
`for loop` 안에선 예외적으로 `i` 라는 줄임말을 쓰는것도 허용됩니다 👍

wheejuni reviewed (approved):
잘 하셨습니다👍 

`groupingBy()` 사용하신 부분도 살펴봤는데, 특별히 어색한 방법이라고 느껴지지 않았습니다.
충분히 잘 구현하신 것 같아서 바로 merge 할게요.

kimnayeon0108 commented:
@wheejuni 감사합니다!!🙇‍♀️
